hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataXceiver.java

```diff
@@ -1078,6 +1079,44 @@ public void blockGroupChecksum(final StripedBlockInfo stripedBlockInfo,
     datanode.metrics.addBlockChecksumOp(elapsed());
   }
 
+  @Override
+  public void copyBlockCrossNamespace(
+      final ExtendedBlock srcBlock, final Token<BlockTokenIdentifier> srcBlockToken,
+      final ExtendedBlock targetBlock, final Token<BlockTokenIdentifier> targetBlockToken,
+      final DatanodeInfo targetDN) throws IOException {
+    updateCurrentThreadName("Copying block cross namespace " + srcBlock);
+    final DataOutputStream reply = getBufferedOutputStream();
+    try {
+      // Check access
+      checkAccess(reply, true, srcBlock, srcBlockToken,
+          Op.COPY_BLOCK_CROSS_NAMESPACE, BlockTokenIdentifier.AccessMode.COPY);
+      checkAccess(reply, true, targetBlock, targetBlockToken,
+          Op.COPY_BLOCK_CROSS_NAMESPACE, BlockTokenIdentifier.AccessMode.WRITE);
+
+      // Async copy the block
+      Future<?> result = datanode.internalCopyBlockCrossNamespace(srcBlock, targetBlock, targetDN);
+      try {
+        result.get(dnConf.getCopyBlockCrossNamespaceSocketTimeout(), TimeUnit.MILLISECONDS);
```

Comment:
If  srcBlock targetBlock not in a same datanode,  it will use  DataTransfer . And DataTransfer will catch any exception in thread , so result.get() will always can not catch IOException.